timesketch/lib/google_jwt.py

```diff
@@ -0,0 +1,166 @@
+"""
+Based on the example code from the public Google IAP documentation:
+https://cloud.google.com/iap/docs/signed-headers-howto
+"""
+
+from __future__ import unicode_literals
+
+import time
+import json
+import jwt
+import requests
+
+from timesketch.lib.definitions import HTTP_STATUS_CODE_OK
+
+
+class JwtValidationError(Exception):
+    """Raised when a JSON Web Token cannot be validated."""
+
+
+class JwtKeyError(Exception):
+    """Raised when there is a problem with the public key used for signing."""
+
+
+def validate_jwt(encoded_jwt, public_key, algorithm, expected_audience,
+                 expected_issuer, expected_domain=None):
+    """Decode anb validate a JSON Web token (JWT).
+
+    Cloud IAP:
+    https://cloud.google.com/iap/docs/signed-headers-howto
+
+    Google OpenID Connect:
+    https://developers.google.com/identity/protocols/OpenIDConnect
+
+    Args:
+        encoded_jwt: The contents of the X-Goog-IAP-JWT-Assertion header.
+        public_key: Key to verify signature of the JWT.
+        algorithm: Algorithm used for the key. E.g. ES256, RS256
+        expected_audience: Expected audience in the JWT.
+        expected_issuer: Expected issuer of the JWT.
+        expected_domain: Expected GSuite domain in the JWT (optional).
+
+    Returns:
+        Decoded JWT on successful validation.
+    """
+    # Decode the token and verify it's payload.
+    try:
+        decoded_jwt = jwt.decode(
+            encoded_jwt, public_key, algorithm=algorithm,
+            audience=expected_audience)
+
+        # Make sure the token is not created in the future or has expired.
+        try:
+            now = int(time.time())
+            issued_at = decoded_jwt['iat']
+            expires_at = decoded_jwt['exp']
+            if issued_at > now:
+                raise JwtValidationError('Token was issued in the future')
+            elif expires_at < now:
+                raise JwtValidationError('Token has expired')
+        except KeyError as e:
+            raise JwtValidationError('Missing timestamp: {}'.format(e))
+
+        # Check that the issuer of the token is correct.
+        try:
+            issuer = decoded_jwt['iss']
+            if issuer != expected_issuer:
+                raise JwtValidationError('Wrong issuer: {}'.format(issuer))
+        except KeyError:
+            raise JwtValidationError('Missing issuer')
+
+        if 'email' not in decoded_jwt:
+            raise JwtValidationError('Missing email field in token')
+
+        if expected_domain:
+            try:
+                domain = decoded_jwt['hd']
+                if not domain == expected_domain:
+                    raise JwtValidationError('Wrong domain: {}'.format(domain))
+            except KeyError as e:
+                raise JwtValidationError('Missing domain: {}'.format(e))
+
+        # If everything checks out, return the decoded token.
+        return decoded_jwt
+
+    except (jwt.exceptions.InvalidTokenError,
```

Comment:
Shouldn't this `except` be just around your statement at line 60? I don't see any other piece of code that could raise this

Reply:
Good catch! Fixed.